src/text-input/index.tsx

```diff
@@ -106,46 +97,81 @@ function formatAutocomplete(autocomplete: string | boolean | undefined): string
 	return autocomplete;
 }
 
-function patternDiffer(
-	previousProperty: string | undefined,
-	newProperty: string | RegExp | undefined
-): PropertyChangeRecord {
-	const value = newProperty instanceof RegExp ? newProperty.source : newProperty;
-	return {
-		changed: previousProperty !== value,
-		value
-	};
-}
+const factory = create({
+	theme,
+	icache,
+	validity,
+	focus,
+	diffProperty,
+	invalidator
+}).properties<TextInputProperties>();
 
-@theme(css)
-@diffProperty('pattern', patternDiffer)
-@diffProperty('leading', reference)
-@diffProperty('trailing', reference)
-export class TextInput extends ThemedMixin(FocusMixin(WidgetBase))<TextInputProperties> {
-	private _onInput(event: Event) {
-		event.stopPropagation();
-		this.properties.onValue &&
-			this.properties.onValue((event.target as HTMLInputElement).value);
-	}
+export const TextInput = factory(function TextInput({
+	middleware: { icache, theme, validity, focus, diffProperty, invalidator },
+	properties
```

Comment:
can you add the `id` here and use it to generate the id base as per https://github.com/dojo/widgets/blob/master/src/menu/index.tsx#L289.
You can then remove `uuid`.

Reply:
Good call!